### ISSUE TITLE
3.0 ORM Using count before all

### ISSUE THREAD

**Issue by Laykou** [cakephp/cakephp#6009]
Using `count()` before `all()` or any iterator doesn't bind associations:

```PHP
// Doesn't work
$query = $articles->contain(['Categories'])->hydrate(false);
$total = $query->count();
foreach ($query as $article) {
  $article['category']; // is empty
}
```

```PHP
// Works
$query = $articles->contain(['Categories'])->hydrate(false);
foreach ($query as $article) {
  $article['category']; // is ok
}
$total = $query->count();
```

Was there recently any change in the `count()` method? The pagination triggers `all()` before `count()` therefore it works properly.

**Comment by lorenzo** [comment]
@markstory it must be the new autoFields thing

**Comment by markstory** [comment]
It very well could be. @Laykou count() recently changed so that in unbinds all the fields from a query to avoid loading additional fields that are not needed for the count. However, that change was done to a 'clean' copy. Perhaps the copy is still sharing an eagerloader instance.

**Comment by mdlouhy** [comment]
I solved this in #6013

**Comment by mdlouhy** [comment]
@Laykou 
This problem occurs when you try to execute a count- and then an all-query with the same instance because when executing count then all associated fields are removed therefore they have to be restored.
```
// this would work because no extra count select would be executed
$query = $articles->contain(['Categories'])->hydrate(false)->all();
$total = $query->count();
foreach ($query as $article) {
  $article['category']; // NOT empty
}
```

**Comment by Laykou** [comment]
Great, thanks @mdlouhy for the fix, I see what was the problem.
@markstory I discovered that it is how PHP works, it only [copies the standard variable types](http://php.net/manual/en/language.oop5.cloning.php):
```
When an object is cloned, PHP 5 will perform a shallow copy of all of the object's properties.
Any properties that are references to other variables, will remain references.
```

Closing this as the #6013 is open. @mdlouhy  please add tests.

**Comment by Laykou** [comment]
Just for the record the issue was solved in #6019
